### PR TITLE
Always include the matching node when resolving point in time 

### LINKED ISSUE #61627: PointInTimeIT#testMultipleIndices failed (Unexpected ShardFailures)
**Build scan**: https://gradle-enterprise.elastic.co/s/alsc5n2wskfyc/console-log?task=:x-pack:plugin:core:internalClusterTest

**Repro line**:

```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:core:internalClusterTest' --tests "org.elasticsearch.xpack.core.search.PointInTimeIT.testMultipleIndices" -Dtests.seed=9D7169CD5CBEFE35 -Dtests.security.manager=true -Dtests.locale=en-MT -Dtests.timezone=Etc/GMT-11 -Druntime.java=11
```

**Reproduces locally?**: Yes (after <5 iterations, doesn't fail reliably)

**Applicable branches**: `master`

**Failure history**: One previous similar-looking failure, in a PR branch: https://gradle-enterprise.elastic.co/s/3fzowvfm3t34i

**Failure excerpt**:

```
org.elasticsearch.xpack.core.search.PointInTimeIT > testMultipleIndices FAILED
    java.lang.AssertionError: Unexpected ShardFailures: [shard [[K3_-IIK9Q_m_2Q3INmxriA][index-2][7]], reason [org.elasticsearch.transport.RemoteTransportException: [node_s1][127.0.0.1:36963][indices:data/read/search[phase/dfs]]
    Caused by: org.elasticsearch.search.SearchContextMissingException: No search context found for id [7]
    	at org.elasticsearch.search.SearchService.findReaderContext(SearchService.java:648)
    	at org.elasticsearch.search.SearchService.getShard(SearchService.java:445)
    	at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:331)
    	at org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$5(SearchTransportService.java:311)
    	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72)
    	at org.elasticsearch.transport.InboundHandler.handleRequest(InboundHandler.java:192)
    	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:99)
    	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:81)
    	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:680)
    	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:142)
    	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:117)
    	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:82)
    	at org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:297)
    	at org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:228)
    	at org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:40)
    	at org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:139)
    	at org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:151)
    	at org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:420)
    	at org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:246)
    	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:174)
    	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:131)
    	at java.lang.Thread.run(Thread.java:834)
    ], cause [org.elasticsearch.search.SearchContextMissingException: No search context found for id [7]
    	at org.elasticsearch.search.SearchService.findReaderContext(SearchService.java:648)
    	at org.elasticsearch.search.SearchService.getShard(SearchService.java:445)
    	at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:331)
    	at org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$5(SearchTransportService.java:311)
    	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72)
    	at org.elasticsearch.transport.InboundHandler.handleRequest(InboundHandler.java:192)
    	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:99)
    	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:81)
    	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:680)
    	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:142)
    	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:117)
    	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:82)
    	at org.elasticsearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:297)
    	at org.elasticsearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:228)
    	at org.elasticsearch.nio.BytesChannelContext.read(BytesChannelContext.java:40)
    	at org.elasticsearch.nio.EventHandler.handleRead(EventHandler.java:139)
    	at org.elasticsearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:151)
    	at org.elasticsearch.nio.NioSelector.handleRead(NioSelector.java:420)
    	at org.elasticsearch.nio.NioSelector.processKey(NioSelector.java:246)
    	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:174)
    	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:131)
    	at java.lang.Thread.run(Thread.java:834)
    ]]
    Expected: <0>
         but: was <1>
        at __randomizedtesting.SeedInfo.seed([9D7169CD5CBEFE35:845FE8F49266CD52]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoFailures(ElasticsearchAssertions.java:305)
        at org.elasticsearch.xpack.core.search.PointInTimeIT.testMultipleIndices(PointInTimeIT.java:137)
REPRODUCE WITH: ./gradlew ':x-pack:plugin:core:internalClusterTest' --tests "org.elasticsearch.xpack.core.search.PointInTimeIT.testMultipleIndices" -Dtests.seed=9D7169CD5CBEFE35 -Dtests.security.manager=true -Dtests.locale=en-MT -Dtests.timezone=Etc/GMT-11 -Druntime.java=11

```

### PR DESCRIPTION
**Pull request by dnhatn**
If shards are relocated to new nodes, then searches with a point in time will fail, although a pit keeps search contexts open. This commit solves this problem by reducing info used by SearchShardIterator and always including the matching nodes when resolving a point in time.

Closes #61627

### CODE REVIEW

**Review by jimczi — COMMENTED** [review]
I don't understand why we need to disable anything since we keep a reference of the index shard in the   local shard reader context ? Are you sure that the problem is the allocation rebalance and not a timeout of the PIT in the test ? If it's not a timeout, we should understand why the reference is not enough when a rebalance occurs. In such case, the PIT should still be valid and the removal of the shard should only happen when the PIT is closed or timed out.

**Comment by dnhatn** [comment]
@jimczi I've updated the PR as we discussed. Can you take another look? Thank you!

**Review by jimczi — APPROVED** [review]
I left one question but the change looks good to me. Thanks for adding the explicit tests for rebalance. That made me think that we should maybe have the same test for deleted indices ?

**Comment by dnhatn** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 1]
```diff
@@ -874,23 +873,28 @@ private static RemoteTransportException wrapRemoteClusterFailure(String clusterA
...
                                                                        SearchContextId searchContext,
                                                                        TimeValue keepAlive) {
         final List<SearchShardIterator> iterators = new ArrayList<>(searchContext.shards().size());
         for (Map.Entry<ShardId, SearchContextIdForNode> entry : searchContext.shards().entrySet()) {
             final ShardId shardId = entry.getKey();
             final ShardIterator shards = OperationRouting.getShards(clusterState, shardId);
-            final List<ShardRouting> matchingNodeFirstRoutings = new ArrayList<>();
+            final List<String> matchingNodeFirst = new ArrayList<>(shards.size());
+            final String nodeId = entry.getValue().getNode();
+            if (clusterState.nodes().get(nodeId) != null) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java]: No, it is not. I pushed https://github.com/elastic/elasticsearch/pull/61658/commits/88072ac6bfc5fb98636c78980d36193378af5a0a

**Comment by dnhatn** [comment]
> That made me think that we should maybe have the same test for deleted indices?

@jimczi Sure, will do. If one index of a PIT is deleted, should we allow search to return a partial result when `allowPartialSearchResults` is true?

**Comment by jimczi** [comment]
> Sure, will do. If one index of a PIT is deleted, should we allow search to return a partial result when allowPartialSearchResults is true?

By deleted, I meant deleted by the user after the creation of the PIT. In such case the PIT should still be valid since we still have the reference of the index shard in the context ? Regarding partial results, yes I think we should allow to return partial results if some contexts are not found. Isn't it what's implemented today ?

**Comment by dnhatn** [comment]
> In such case the PIT should still be valid since we still have the reference of the index shard in the context ?

I think we [remove](https://github.com/elastic/elasticsearch/blob/a70c00a62cf8eceafb93ef7c2fe3aee4c6930de9/server/src/main/java/org/elasticsearch/search/SearchService.java#L290-L292) all reader contexts when an index is deleted.

> Regarding partial results, yes I think we should allow to return partial results if some contexts are not found. Isn't it what's implemented today ?

We don't. I will work on it in a follow-up.

**Pull request merged**